wagtail/contrib/redirects/tests.py

```diff
@@ -119,7 +119,7 @@ def test_temporary_redirect(self):
         # Check that we were redirected temporarily
         self.assertRedirects(response, '/redirectto', status_code=302, fetch_redirect_response=False)
 
-    def test_redirect_stripping_query_string(self):
+    def test_redirect_matches_query_string(self):
```

Comment:
Often, sites are migrated from a platform where the page is identified by a query parameter. So a user might add the following redirect:

``/index.php?page_id=1`` to ``/about-us/``

But it seems the logic here will actually redirect to ``/about-us/?page_id=1``.

Is it worth adding logic in to handle this?

Also, ideally, ``/index.php?page_id=1&utm_source=foo`` should forward to ``/about-us/?utm_source=foo``

Reply:
how would we know which querystring parameters to keep and which ones to throw away?

Reply:
I'm thinking that any query parameters that were specified in the "old path" field should be thrown away, unless also specified in the redirect field. Everything else should be passed through.

This logic might be a little confusing to users though...

Reply:
Regarding the discussed options, I think there are two actual use-cases:
1. You have rounded up a migration to Wagtail, therefore `/index.php?page-id=5` should redirect to `/about-us/`.
2. You are currently changing URLs in your Wagtail site, therefore `/previous-search-page/?q=keyword` should redirect to `/new-search-page/?q=keyword`.

The edge-case would be `/index.php?page-id=5&utm_source=12` but supporting that would be quite a hassle and would be really confusing for the end-user to configure.

To make it feasible for both scenarios it could make sense to add a checkbox whether the editor wants to preserve querystrings while configuring the redirect, the same way it asks whether it should permanently redirect.  

Curious to see both your perspectives on this.

Reply:
@jjanssen I think it'd be possible to handle the edge case you mentioned without too much additional craziness.

The rules in english would be:

* Specified source params are stripped if not present in the target
* Unspecified params are passed through

```
from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5
redirect    /about-us/

from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5&utm_source=12
redirect    /about-us/?utm_source=12

from        /previous-search-page/
to          /new-search-page/
visit       /previous-search-page/?q=keyword
redirect    /new-search-page/?q=keyword

from        /previous-search-page/?q=keyword
to          /new-search-page/?q=keyword
visit       /previous-search-page/?q=keyword
redirect    /new-search-page/?q=keyword
```

what do you think?

Reply:
@MattSegal 👍 That sounds like the ideal behaviour to me.

Reply:
Sounds like a great idea Matt!

Reply:
Just pushed some new changes. Turns out I couldn't do everything I said that I would do. Most of it, but not everything. I think/hope my tests make the expected behavior pretty clear.

## Can do:

Pass through requested params

```
from        /previous-search-page
to          /new-search-page
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

Strip old params:

```
from        /index.php?page-id=5
to          /about-us
visit       /index.php?page-id=5
redirect    /about-us
```

Merge requested and specified params:

```
from        /index.php
to          /about-us?foo=1
visit       /index.php?utm_source=12
redirect    /about-us?utm_source=12&foo=1
```

Respect redirected params

```
from        /previous-search-page?q=keyword
to          /new-search-page?q=keyword
visit       /previous-search-page?q=keyword
redirect    /new-search-page?q=keyword
```

## Can't do:

If you visit a URL that doesn't match the old_path querystring exactly, then we won't be able to find a redirect. Even if we could there are potential conflicts where 2 redirects could match 1 request. As such, we can't easily do this:

```
from        /index.php?page-id=5
to          /about-us/
visit       /index.php?page-id=5&utm_source=12
redirect    /about-us/?utm_source=12
```

Instead we have to 404, or look for a redirect that doesn't have a querystring.